clang-tools: move into llvmPackages

ShamrockLee opened the pull request:
###### Description of changes

Format the Nix expression of `clang-tools` using `nixfmt`, in accordance with NixOS/rfcs#166.

Move the The clang-tools expression folder under `llvm/common` and `clang-tools` package under `llvmPackages`.

Move `clang-tools_<version>` into `aliases.nix`, specifying as `llvmPackages_<version>.clang-tools`.

~Provide `clang-tools-python` that links and wraps executables from `clang-unwrapped.python` output. If applied, users will be able to enjoy `git clang-format` with `clang-tools-python` installed (or introduced inside Nix shell).~

~As for `"${clang-unwrapped.python}/bin/scan-view"`, it depends on the relative module inside `"../share"`. The dependent module of `scan-view`, `ScanView.py`, is currently not moved into `clang-unwrapped.python`, which will be fixed by the staging PR #191801.~

Cc:
@Patryk27 clang-tools maintainer

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage) (The 24 updated packages will take some time to build.)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`) (Tested git-clang-format, but not scan-view)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ShamrockLee commented:
@ofborg build bat-extras.prettybat cloudcompare entwine grass libpulsar nominatim pdal python310Packages.tiledb python39Packages.tiledb qgis qgis-ltr tiledb vscode-extensions.ms-vscode-cpptools

ShamrockLee commented:
Just seperate those Python scripts to a new output `python`, so that packages depending on clang-tools won't be forced to have Python in their closure.

Patryk27 reviewed (approved):
Looks pretty neat 🙂

ShamrockLee commented:
Could it be merged now?

Or should I split the wrapped python script into a new derivation `clang-tools-python` to reduce the number of rebuild?

Patryk27 commented:
The changes look fine to me -- I don't personally have merging rights, though 👀

ShamrockLee commented:
No wonder. I had thought that all the members have the merging right.

ShamrockLee commented:
Just make a new derivation `clang-tools-python` instead. This way, packages depending on `clang-tools` won't have to rebuild whenever python gets rebuild.

Wrapping the Python scripts into a new package also guarantees zero rebuild of the current packages, and is thus easier to backport.

Patryk27 reviewed (approved):
🚀

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1507

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13660,6 +13660,50 @@ with pkgs;
     llvmPackages = llvmPackages_14;
   };
 
+  clang-tools-python = callPackage ../development/tools/clang-tools/python.nix {
+    llvmPackages = llvmPackages_latest;
+  };
+
+  clang-tools-python_5 = callPackage ../development/tools/clang-tools/python.nix {
+    llvmPackages = llvmPackages_5;
+  };
+
+  clang-tools-python_6 = callPackage ../development/tools/clang-tools/python.nix {
+    llvmPackages = llvmPackages_6;
+  };
+
+  clang-tools-python_7 = callPackage ../development/tools/clang-tools/python.nix {
+    llvmPackages = llvmPackages_7;
+  };
+
+  clang-tools-python_8 = callPackage ../development/tools/clang-tools/python.nix {
+    llvmPackages = llvmPackages_8;
```
Why don't we move it under llvmPackages?

ShamrockLee replied:
Sounds great.

Not sure what would be the proper way to move it into `llvmPackages`. It's quite confusing to see such one-version-per-folder form of packaging.

ShamrockLee replied:
Just moved `clang-tools` and `clang-tools-python` into `llvmPackages`.

ShamrockLee replied:
Update: Strange errors occurred.

ShamrockLee replied:
Addressed.

SuperSandro2000 commented on pkgs/development/tools/clang-tools/python.nix:
```diff
@@ -0,0 +1,60 @@
+  '';
+
+  meta = clang-unwrapped.meta // {
+    description = "Standalone Python scripts provided by Clang";
```
What about homepage?

ShamrockLee replied:
The `homepage` attribute is inherited from `clang-unwrapped.meta`.

`clang-tools` and `clang-tools-python` are merely wrappers to the corresponding files provided by the `clang-unwrapped` package, so the homepage should be the same.

ShamrockLee replied:
Update: I have changed my mind after seeing `meta.name` and `meta.mainProgram` from clang-unwrapped.meta.name`.

The attributes to inherit should be explicitly specified as you suggested for the `raspa` PR (https://github.com/NixOS/nixpkgs/pull/186127#discussion_r1039016406).

SuperSandro2000 commented on pkgs/development/tools/clang-tools/python.nix:
```diff
@@ -0,0 +1,60 @@
+  ];
+
+  clang_unwrapped_python = clang-unwrapped.python;
+  pythonInterpreter = python3.interpreter;
```
That is unused I think

SuperSandro2000 commented on pkgs/development/tools/clang-tools/python.nix:
```diff
@@ -0,0 +1,60 @@
+    xorg.lndir
+  ];
+
+  clang_unwrapped_python = clang-unwrapped.python;
```
and this should be inline

ShamrockLee commented:
@ofborg build clang-tools-python clang-tools-python.passthru.tests

ShamrockLee commented:
@SuperSandro2000 I don't know why, but it keeps complaining

```
error: Function called without required argument "clang-tools-extra_src" at /run/media/root/data-btrfs/shamrock-shared/targets/Projects/NixOS/nixpkgs_optparse/pkgs/development/compilers/llvm/5/clang/default.nix:1
```
and
```
error: Function called without required argument "llvm_meta" at /run/media/root/data-btrfs/shamrock-shared/targets/Projects/NixOS/nixpkgs_optparse/pkgs/development/compilers/llvm/14/clang/default.nix:1
```

I just specify
```
clang-tools = callPackage ../common/clang-tools {
  inherit (tools) clang-unwrapped;
};
```
as other packages in llvmPackages_5 does.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/57

ShamrockLee commented:
Issues solved!

ShamrockLee commented:
Do we consider

1. Changing the `<pkg>.override` interface
2. Moving packages into `aliases.nix`

breaking changes?

ShamrockLee commented:
Refine the package test.

ShamrockLee commented:
Bump onto master and resolve merge conflict.

ShamrockLee commented:
I'm splitting this PR to preserve only the `clang-tools`'s movig to `llvmPackages` part. This is a backward-incompatible change, and I hope it could go into Nixpkgs 24.05.

This also allows me to re-think if `clang-tools-python` is the suitable way to provide `clang-tools-python`. I just noticed the existence of `clang-analyzer`, which provides `scan-build` and `scan-view`. After removing `scan-view` from `clang-tools-python`, all it provides will be `git-clang-format`.

Patryk27 commented:
Hi, sure! I'll take a look tomorrow 🙂

ShamrockLee committed:
clang-tools: format with nixfmt (Nix RFC 166)

Format default.nix with nixfmt in accordance with Nix RFC 166.

Manually Place the comments above the corresponding argument.

ShamrockLee committed:
clang-tools: move into llvmPackages

ShamrockLee committed:
clang-tools_*: move to aliases

ShamrockLee committed:
clang-tools: add release note entry about moving into llvmPackages

Add 24.11 release note entry about moving clang-tools into llvmPackages
and making clang-tools_<version> aliases.

ShamrockLee commented:
I bumped it onto the `master` branch, resolved merge conflicts, and revised the release note to use imperative tones.

RossComputerGuy reviewed (approved):
Just resolve the conflicts and this should be good to go.

ShamrockLee commented:
@RossComputerGuy The release note entry is inside `rl-2405.section.md`. Is it okay to backport this PR, or how should I handle the release note entry?

RossComputerGuy commented:
@ShamrockLee I don't think it's too late to backport.

rrbutani commented:
> Is it okay to backport this PR

> I don't think it's too late to backport

I think we should maybe get a second opinion on this; breaking changes are restricted <4 weeks before [24.05's release](https://github.com/NixOS/nixpkgs/issues/303285 and changing the `clang-tools_*` attributes to aliases could be construed as breaking.

> Imo, release note entries are not that important for something like this.

I think it's worth having, especially since the intent (as per the OP) is to phase out the aliases that were added.

ShamrockLee commented:
@rrbutani Not sure if the change of `<pkg>.override` interfaces caused by moving `clang-tools` into `llvmPakcages` would also be considered backward-incompatible.

rrbutani reviewed (changes requested):
Looks good, couple of nits about the release note entry:

rrbutani commented on pkgs/development/compilers/llvm/common/clang-tools/default.nix:
```diff
@@ -1,17 +1,22 @@
 
   pname = "clang-tools";
-  version = lib.getVersion unwrapped;
+  version = lib.getVersion clang-unwrapped;
```
nit: the scope that's used in `llvmPackages` already has `version`:

https://github.com/NixOS/nixpkgs/blob/0179f8c547879eb26016fb1dcda2904001487a3d/pkgs/development/compilers/llvm/18/default.nix#L61

You should be able to drop `clang-unwrapped` from this package's attrs (and all of the `llvmPackages_*` set's `default.nix`s and just ask for `version` in this package's attrs.

ShamrockLee replied:
Since `clang-tools` is a wrapper of (a part of) `clang-unwrapped`, the build process would be more straightforward to pass `clang-unwrapped` as a shell variable. It's also okay to string-interpolate it directly.

As for the version, it makes sense to pass it from the llvmPackage an `<pkg>.override` argument.

> You should be able to drop clang-unwrapped from this package's attrs

Are we removing the attribute to prevent people from accessing the `clang-unwrapped` package used via the attribute (or even try to override it via `<pkg>.overrideAttrs`) but encourage accessing/overriding via `llvmPackages` instead?

rrbutani replied:
My apologies, I missed that the `$out/bin` symlinks are produced from `clang-unwrapped` and not `clang`/`libcxxClang`.

Please disregard the above; getting `version` from `clang-unwrapped` seems like the sensible thing to do.

ShamrockLee commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -563,6 +563,12 @@ can now be used to disable the upstream fallback servers entirely by setting it
 
 - `virtialisation.incus` now defaults to the newly-added `incus-lts` release (v6.0.x). Users who wish to continue using the non-LTS release will need to set `virtualisation.incus.package = pkgs.incus`. Stable release users are encouraged to stay on the LTS release as non-LTS releases will by default not be backported.
 
+- Move `clang-tools` into `llvmPackages`.
```
Addressed. Thank you for pointing out.

By the way, I once received a review from @fricklerhandwerk in another PR that suggests using imperative tones when writing release notes. However, most of the entries in this edited release note use passive tones so that the target of changes becomes the first word of a sentence.

Do you think I should change the newly added entry and use a passive tone instead?

rrbutani replied:
As far as I know indicative mood with the subject up-front is the aim here but @fricklerhandwerk likely knows better than I do; I don't interact with the release notes much 🙂 

cc: @jacg

jacg replied:
TLDR: yes, please change the entry to match the style used throughout the rest of document.

For the 24.05 release, I was instructed to ensure that the name of the affected item appear at or near the beginning of every note (in order to make it easier to spot relevant entries when scanning the document), hence the subject-up-front indicative style is used throughout the document. This was already the prevailing style before I started editing the notes.

I would therefore argue that, in this context, all notes should follow that style, regardless of what merits other styles might have in other contexts.

rrbutani commented:
> Not sure if the change of `<pkg>.override` interfaces caused by moving `clang-tools` into `llvmPakcages` would also be considered backward-incompatible.

oh, good point! here, let's ask:

---

@wegank This PR changes some top-level attrs to be aliases and changes some package attributes such that existing `clang-tools_<ver>.override { ... }` invocations may break (now takes `clang`, `libcxxClang` instead of `llvmPackages`). We're wondering if this is okay to backport to 24.05 or if this is considered a breaking change.

wegank commented:
I'd consider the change of argument for `clang-tools` to be breaking. There's also the move to aliases for `clang-tools_1{2-8}`, which sounds breaking too.

ShamrockLee commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -563,6 +563,12 @@ can now be used to disable the upstream fallback servers entirely by setting it
+  Alias `clang-tools_<version>` to `llvmPackages_<version>.clang-tools`; remove in future releases.
+
+  For users' convenience, keep the top-level attribute `clang-tools` and specify it as `llvmPackages.clang-tools`.
+
```
Addressed.

rrbutani commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -234,6 +234,10 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
   To be able to play DRM protected content in `chromium` now, you have to explicitly opt-in as originally intended using `chromium.override { enableWideVine = true; }`.
   This override was added almost 10 years ago.
 
+- `clang-tools_<version>` packages have been moved into `llvmPackages_<version>` (i.e. `clang-tools_18` is now `llvmPackages_18.clang-tools`).
+  - For convenience, the top-level `clang-tools` attribute remains and is now bound to `llvmPackages.clang-tools`.
+  - Top-level `clang_tools_<version>` attributes are now aliases; these will be removed in a future release.
+
```
Since we're not backporting this PR we should move this to the [24.11 release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md#backward-incompatibilities-sec-release-2411-incompatibilities).

ShamrockLee replied:
Addressed.

rrbutani reviewed (changes requested):
@ShamrockLee thanks for working on this; one last minor change and this will be good to merge:

ShamrockLee commented:
I just rebased the feature branch on top of the `master` branch and resolved the merge conflict of the release note entries.

Is there anything I could help to push it forward?

ShamrockLee commented:
I simplified the package calling of `clang-tools` in each `llvmPackages` and fixed the evaluation, i.e.

```diff
     # Wrapper for standalone command line utilities
-    clang-tools = callPackage ../common/clang-tools {
-      inherit (tools) clang-unwrapped clang libcxxClang;
-      inherit llvm_meta;
-    };
+    clang-tools = callPackage ../common/clang-tools { };
```

JohnRTitor commented:
Hi, could you rebase and remove the conflicts please? I'll merge it straight after that.

JohnRTitor commented:
Result of `nixpkgs-review pr 191698` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>llvmPackages_12.clang-tools</li>
    <li>llvmPackages_13.clang-tools</li>
    <li>llvmPackages_14.clang-tools</li>
    <li>llvmPackages_15.clang-tools</li>
    <li>llvmPackages_16.clang-tools</li>
    <li>llvmPackages_18.clang-tools</li>
  </ul>
</details>
